File: 백준/Gold/11444. 피보나치 수 6/피보나치 수 6.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

n = int(input())
matrix = [[1, 1], [1, 0]]

# 행렬 곱셈
def mul_matrix(matrix1, matrix2):
    res = [[0]*2 for _ in range(2)]
    for i in range(2):
        for j in range(2):
            for k in range(2):
                res[i][j] += matrix1[i][k] * matrix2[k][j]% 1000000007
    return res

def power(a, b):
    if b == 1:  # b의 값이 1이 될 때까지 재귀
        return a
    else:
        tmp = power(a, b // 2)  # a^(b // 2)
        if b % 2 == 0:
            return mul_matrix(tmp, tmp)  # b%2==0
        else:
            return mul_matrix(mul_matrix(tmp, tmp), a)  #else
result = power(matrix, n)

print(result[0][1]% 1000000007)

File: 백준/Gold/11444. 피보나치 수 6/피보나치 수 6.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	matrix = [][]int{}
)

func mulMatrix(matrix1, matrix2 [][]int) [][]int {
	res := [][]int{}
	res = append(res, []int{0, 0})
	res = append(res, []int{0, 0})
	for i := 0; i < 2; i++ {
		for j := 0; j < 2; j++ {
			for k := 0; k < 2; k++ {
				res[i][j] += matrix1[i][k] * matrix2[k][j] % 1000000007
			}
		}
	}
	return res
}

func power(a [][]int, b uint64) [][]int {
	if b == 1 {
		return a
	}
	tmp := power(a, b/2)
	if b%2 == 0 {
		return mulMatrix(tmp, tmp)
	} else {
		return mulMatrix(mulMatrix(tmp, tmp), a)
	}
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	var n uint64
	matrix = append(matrix, []int{1, 1})
	matrix = append(matrix, []int{1, 0})
	fmt.Fscanf(reader, "%d\n", &n)
	result := power(matrix, n)
	fmt.Fprintf(writer, "%d", result[0][1]%1000000007)
}
